File: VB/XPO/CustomEventsAndResources.Module/BusinessObjects/CustomResource.vb
Imports System.ComponentModel
Imports System.Drawing
Imports DevExpress.Persistent.Base.General
Imports DevExpress.Xpo
Imports DomainComponents.Common
Imports DevExpress.Persistent.BaseImpl

Namespace CustomEventsAndResources.Module.BusinessObjects

    '[DefaultProperty(nameof(Caption))]
    Public Class CustomResource
        Inherits BaseObject
        Implements IResource

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub

        Public Overrides Sub AfterConstruction()
            MyBase.AfterConstruction()
            Color = Color.White
        End Sub

        <Browsable(False)>
        Public ReadOnly Property Id As Object Implements IResource.Id
            Get
                Return Me.Oid
            End Get
        End Property

        Public Property Caption As String Implements IResource.Caption
            Get
                Return GetPropertyValue(Of String)(NameOf(CustomResource.Caption))
            End Get

            Set(ByVal value As String)
                SetPropertyValue(NameOf(CustomResource.Caption), value)
            End Set
        End Property

        <Browsable(False)>
        Public ReadOnly Property OleColor As Int32
            Get
                Return ColorTranslator.ToOle(Color)
            End Get
        End Property

        <Association("CustomEventWithCustomResource-CustomResource")>
        Public ReadOnly Property Events As XPCollection(Of CustomEventWithCustomResource)
            Get
                Return GetCollection(Of CustomEventWithCustomResource)(NameOf(CustomResource.Events))
            End Get
        End Property

        <ValueConverter(GetType(ColorToIntConverter))>
        Public Property Color As Color
            Get
                Return GetPropertyValue(Of Color)(NameOf(CustomResource.Color))
            End Get

            Set(ByVal value As Color)
                SetPropertyValue(NameOf(CustomResource.Color), value)
            End Set
        End Property
    End Class
End Namespace

File: CS/XPO/CustomEventsAndResources.Module/BusinessObjects/CustomResource.cs
using System.ComponentModel;
using System.Drawing;
using DevExpress.Persistent.Base.General;
using DevExpress.Xpo;
using DomainComponents.Common;
using DevExpress.Persistent.BaseImpl;

namespace CustomEventsAndResources.Module.BusinessObjects; 
//[DefaultProperty(nameof(Caption))]
public class CustomResource : BaseObject, IResource {
    public CustomResource(Session session) : base(session) { }
    public override void AfterConstruction() {
        base.AfterConstruction();
        Color = Color.White;
    }
    [Browsable(false)]
    public object Id {
        get => Oid;
    }
    public string Caption {
        get => GetPropertyValue<string>(nameof(Caption));
        set => SetPropertyValue(nameof(Caption), value);
    }
    [Browsable(false)]
    public Int32 OleColor {
        get { return ColorTranslator.ToOle(Color); }
    }
    [Association("CustomEventWithCustomResource-CustomResource")]
    public XPCollection<CustomEventWithCustomResource> Events {
        get { return GetCollection<CustomEventWithCustomResource>(nameof(Events)); }
    }
    [ValueConverter(typeof(ColorToIntConverter))]
    public Color Color {
        get => GetPropertyValue<Color>(nameof(Color));
        set => SetPropertyValue(nameof(Color), value);
    }
}
